### ISSUE TITLE
Nix now runs out of memory

### ISSUE THREAD

**Issue by drewr** [commercialhaskell/stack#4010]
A stack script I've run for a while now with the following head:

```
#!/usr/bin/env stack
{- stack
    --resolver=lts-9.21
    --nix
    --no-nix-pure
    runghc
    --package filemanip
    --package HStringTemplate
    --package regex-compat
    --package shake
    --package split
-}
```
Now fails to run when it actually needs to install a bunch of deps from Nix:
```
[...]
  /nix/store/rkz6ajid7v27nkr6n223yhx9kbd7c4ql-perl-IO-HTML-1.001
  /nix/store/shhkmb9rsf3q2rxaw0jban9jqbinh0v6-perl-LWP-MediaTypes-6.02
  /nix/store/vs091h9m987lcdqili0qmxchk1pw5329-zlib-1.2.11-dev
  /nix/store/vx7x6j9qy1mxj6b8p7143yfkp2wipwfq-perl-HTML-Parser-3.72
  /nix/store/zxcg1wqr18qz42m9jx9dixzfx558nl8c-perl-HTTP-Message-6.16
copying path '/nix/store/12s27v3lvvg8dy43gnrwlykn2n6i4i1r-gmp-6.1.2-info' from 'https://cache.nixos.org'...
copying path '/nix/store/35v3v6rc0vcichqs8dsxg5h8blanj3kq-ghc-8.0.2' from 'https://cache.nixos.org'...
error: out of memory
make: *** [Makefile:8: test] Error 1
```

I'm using a `stack` binary from running `stack upgrade`:
```
% which stack
~/.local/bin/stack
% stack --version
Version 1.7.1, Git revision 681c800873816c022739ca7ed14755e85a579565 (5807 commits) x86_64 hpack-0.28.2
```

Also, Nix is latest:
```
% nix --version    
nix (Nix) 2.0.1
```

The Nix author seems to think this is expected (https://github.com/NixOS/nixpkgs/issues/38635). If that's the case, is there a workaround in stack?

**Comment by nh2** [comment]
I think you're observing a regression of Nix 2.0 that copied package contents into memory instead of streaming them to disk (thus it failing for a large package like ghc): https://github.com/NixOS/nix/issues/1681

Most of these issues are fixed in nix `master`, and likely to be released with nix 2.0.5.

The last known issue might be fixed when https://github.com/NixOS/nix/pull/2206 is merged, but maybe that one isn't necessary for your use case.

So I recommend to either

* hold out until nix 2.0.5 is released, or
* use nix `master` (perhaps with that PR cherry-picked), or
* use nix < 2.0

**Comment by drewr** [comment]
I figured there was some kind of regression somewhere. Those workarounds are reasonable. Thank you @nh2 for making sense of things!
